cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -0,0 +1,379 @@
+    while (col.type().id() == type_id::LIST or col.type().id() == type_id::STRUCT) {
+      if (col.type().id() == type_id::STRUCT) {
+        start_value_idx += col.offset();
+        end_value_idx += col.offset();
+        col = col.child(0);
+      } else {
+        auto offset_col = col.child(lists_column_view::offsets_column_index);
+        start_value_idx = offset_col.element<size_type>(start_value_idx + col.offset());
+        end_value_idx   = offset_col.element<size_type>(end_value_idx + col.offset());
+        col             = col.child(lists_column_view::child_column_index);
+      }
+    }
+    s_start_value_idx = start_value_idx;
+    s_num_values      = end_value_idx - start_value_idx;
+  }
+  __syncthreads();
+
+  column_device_view &data_col = *s_col.leaf_column;
+  using block_reduce           = cub::BlockReduce<size_type, block_size>;
+  __shared__ typename block_reduce::TempStorage reduce_storage;
+
+  // Make a view of the hash map
+  auto hash_map_mutable = map_type::device_mutable_view(
+    s_chunk->dict_map_slots, s_chunk->dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
+  auto hash_map = map_type::device_view(
+    s_chunk->dict_map_slots, s_chunk->dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
+
+  for (size_type i = 0; i < s_num_values; i += block_size) {
+    // Check if the num unique values in chunk has already exceeded max dict size and early exit
+    if (s_chunk->num_dict_entries > MAX_DICT_SIZE) { return; }
+
+    // add the value to hash map
+    size_type val_idx = i + t + s_start_value_idx;
+    bool is_valid =
+      (i + t < s_num_values && val_idx < data_col.size()) ? data_col.is_valid(val_idx) : false;
+
+    // insert element at val_idx to hash map and count successful insertions
+    size_type is_unique      = 0;
+    size_type uniq_elem_size = 0;
+    if (is_valid) {
+      auto found_slot = type_dispatcher(data_col.type(), map_find_fn{hash_map}, data_col, val_idx);
+      if (found_slot == hash_map.end()) {
+        is_unique =
+          type_dispatcher(data_col.type(), map_insert_fn{hash_map_mutable}, data_col, val_idx);
+        uniq_elem_size = [&]() -> size_type {
+          if (not is_unique) { return 0; }
+          switch (s_col.physical_type) {
+            case Type::INT32: return 4;
+            case Type::INT64: return 8;
+            case Type::INT96: return 12;
+            case Type::FLOAT: return 4;
+            case Type::DOUBLE: return 8;
+            case Type::BYTE_ARRAY:
+              if (data_col.type().id() == type_id::STRING) {
+                // Strings are stored as 4 byte length + string bytes
+                return 4 + data_col.element<string_view>(val_idx).size_bytes();
+              }
+            case Type::FIXED_LEN_BYTE_ARRAY:
+            default: cudf_assert(false && "Unsupported type for dictionary encoding"); return 0;
+          }
+        }();
+      }
+    }
+
+    // TODO: Explore using ballot and popc for this reduction as the value to reduce is 1 or 0
```

Comment:
Still have "TODO"? Something is not done yet?

Reply:
This would've been premature optimization. Since all values are 1 or 0, we could use an intrinsic to calculate it faster. However it takes such a small part of the execution time that adding the extra code is not worth it and a potential for errors.